src/main/java/org/candlepin/model/OwnerCurator.java

```diff
@@ -199,45 +207,59 @@ public Owner getByUpstreamUuid(String upstreamUuid) {
      *  a list of owners associated with the specified products
      */
     @SuppressWarnings("checkstyle:indentation")
-    public CandlepinQuery<Owner> getOwnersWithProducts(Collection<String> productIds) {
-        if (productIds != null && !productIds.isEmpty()) {
-            Session session = this.currentSession();
-
-            // Impl note:
-            // We have to break this up into two queries for proper cursor and pagination support.
-            // Hibernate currently has two nasty "features" which break these in their own special
-            // way:
-            // - Distinct, when applied in any way outside of direct SQL, happens in Hibernate
-            //   *after* the results are pulled down, if and only if the results are fetched as a
-            //   list. The filtering does not happen when the results are fetched with a cursor.
-            // - Because result limiting (first+last result specifications) happens at the query
-            //   level and distinct filtering does not, cursor-based pagination breaks due to
-            //   potential results being removed after a page of results is fetched.
-            Criteria idCriteria = session.createCriteria(Owner.class, "Owner")
-                .setProjection(Projections.distinct(Projections.id()))
-                .createAlias("Owner.pools", "Pool")
-                .createAlias("Pool.product", "Prod", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("Prod.derivedProduct", "DProd", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("Prod.providedProducts", "PProd", JoinType.LEFT_OUTER_JOIN)
-                .createAlias("DProd.providedProducts", "DPProd", JoinType.LEFT_OUTER_JOIN)
-                .add(Restrictions.or(
-                    CPRestrictions.in("Prod.id", productIds),
-                    CPRestrictions.in("DProd.id", productIds),
-                    CPRestrictions.in("PProd.id", productIds),
-                    CPRestrictions.in("DPProd.id", productIds)
-                ));
-
-            List<String> ownerIds = idCriteria.list();
-
-            if (ownerIds != null && !ownerIds.isEmpty()) {
-                DetachedCriteria criteria = DetachedCriteria.forClass(Owner.class)
-                    .add(CPRestrictions.in("id", ownerIds));
-
-                return this.cpQueryFactory.buildQuery(session, criteria);
+    public List<Owner> getOwnersWithProducts(Collection<String> productIds) {
+        if (productIds == null || productIds.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        CriteriaBuilder criteriaBuilder = this.entityManager.get().getCriteriaBuilder();
+        CriteriaQuery<String> query = criteriaBuilder.createQuery(String.class);
+        Root<Owner> owner = query.from(Owner.class);
+        query.distinct(true);
```

Comment:
With these updated queries, distinct restriction is applied on the SQL query directly (verified from the debug logs). The first query fetches the distinct owner IDs & second query is for fetching the owner entities. Should we merge these two into a single query?

Since we are not doing any pagination, we are safe from the hibernate returning incorrect data when we deal with fetching distinct IDs with joins & pagination.  
I have added the comment/note back, explaining the issue with pagination for future reference.

Reply:
If DISTINCT is being applied properly, then the original issue is resolved and we can combine the two queries back into one -- especially since we're partitioning the queries and collecting the inputs already.

Another thing to consider here is that pagination is busted as soon as we break the query up at all (as we're doing with the partitioning); we need to either sort the product ID collection before we partition, or sort the output list after we run through all of the blocks.

Reply:
@Ceiu @nikosmoum 
The original return type for query was ```CandlepinQuery``` which means pagination need to be supported with updated queries as well if someone tries to access endpoint with PageRequest. 
To support pagination we need fetch the total number of records, which is to be set as ```maxRecords``` in PageRequest. Here instead of getting the count, we fetch distinct OwnerIds (with JPA the distinct keyword is the part of SQL query) & use this in second query to fetch the owner entities.

With updated queries there could be one potential issue [here](https://github.com/candlepin/candlepin/blob/6e99912c80b1b549dbdceb5802bd92ff3a21bad9/src/main/java/org/candlepin/model/OwnerCurator.java#L231) as in order to support the ```sort_by``` functionality of pagination, we need to pass all the ownerIDs to IN clause by using partition which could still exceed the parameter limits. One solution here could be to break the second query again to fetch sub-set of a page, and later merge the results into single page. What are you suggestions ?

Reply:
Thanks for detailed explanation, since endpoint was never intended to explicitly support paging, we are okay to not support it. I can check with Katello side to verify that they are not calling API with such request. For now I have removed the paging bits, & used simple partition technique to get the work done.

Reply:
We need to verify with both Katello and Teamnado that they're not using pagination. If anyone was using pagination, I'd expect that to be Teamndo

Reply:
@nikosmoum I found Katello is not using this API in their codebase as of now. Teamnado have used this endpoint (without pagination) in the past & actively wants to use this for errata notification service, they don't require this API to be paginated as they didn't face any issues in the past.